File: node-js/oauth.js

require('dotenv').config();
const {CLIENT_ID, CLIENT_SECRET, TOKEN_HOST, TOKEN_PATH, SCOPE} = process.env

const config = {
  client: {
    id: CLIENT_ID,
    secret:
      CLIENT_SECRET,
  },
  auth: {
    tokenHost: TOKEN_HOST,
    tokenPath: TOKEN_PATH
  },
};
const { ClientCredentials } = require("simple-oauth2");
const client = new ClientCredentials(config);

const tokenParams = {
  scope: SCOPE
};

async function generateAccessToken() {
  try {
    return await client.getToken(tokenParams);
  } catch (error) {
    console.error("Error gathering access token: ", error.message);
  }
}

async function refreshTokenIfExpired(accessToken) {
  if (accessToken.expired()) {
    console.info("Refreshing token....");
    return generateAccessToken();
  } else {
    console.info("Token not expired, returning original...");
    return accessToken;
  }
}

async function run() {
  const accessToken = await generateAccessToken();

  console.log(accessToken);
  const refreshToken = await refreshTokenIfExpired(accessToken);

  console.log(refreshToken);
}


module.exports = {
  generateAccessToken,
  refreshTokenIfExpired,
  run
};

File: go/oauth.go
package main

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"time"

	"github.com/joho/godotenv"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/clientcredentials"
)

func main() {
	// Load environment variables
	err := godotenv.Load()
	clientID := os.Getenv("CLIENT_ID")
	clientSecret := os.Getenv("CLIENT_SECRET")
	tokenHost := os.Getenv("TOKEN_HOST")
	tokenPath := os.Getenv("TOKEN_PATH")
	scope := os.Getenv("SCOPE")

	// Configuration
	config := &clientcredentials.Config{
		ClientID:     clientID,
		ClientSecret: clientSecret,
		TokenURL:     fmt.Sprintf("%s%s", tokenHost, tokenPath),
		Scopes:       []string{scope},
	}
	// Create OAuth2 client
	client := config.Client(context.Background())

	// Run the functions
	accessToken, err := generateAccessToken(client)
	if err != nil {
		fmt.Printf("Error gathering access token: %s\n", err.Error())
		return
	}

	fmt.Printf("Access Token: %s\n", accessToken)

	refreshToken, err := refreshTokenIfExpired(config, accessToken)
	if err != nil {
		fmt.Printf("Error refreshing token: %s\n", err.Error())
		return
	}

	fmt.Printf("Refreshed Token: %s\n", refreshToken)

}

func generateAccessToken(client *http.Client) (string, error) {
	// Obtain access token
	token, err := client.Transport.(*oauth2.Transport).Source.Token()
	fmt.Println(token)
	if err != nil {
		return "", err
	}
	return token.AccessToken, nil
}

func refreshTokenIfExpired(config *clientcredentials.Config, accessToken string) (string, error) {
	// Check if the access token is expired
	currentTime := int(time.Now().Unix())
	tokenInfo, err := config.Token(context.Background())
	if err != nil {
		return "", err
	}

	expirationTime := int(tokenInfo.Expiry.Unix())
	if expirationTime <= currentTime {
		fmt.Println("Token is expired. Refreshing...")
		accessToken, err := generateAccessToken(config.Client(context.Background()))
		if err != nil {
			return "", err
		}
		return accessToken, nil
	} else {
		fmt.Println("Token is not expired. Returning original")
		return accessToken, nil
	}
}
